usvm-ts/src/main/kotlin/org/usvm/machine/operator/TsBinaryOperator.kt

```diff
@@ -488,46 +525,47 @@ sealed interface TsBinaryOperator {
         ): UBoolExpr {
             check(lhs.isFakeObject() || rhs.isFakeObject())
 
-            when {
+            val typeConstraint = when {
                 lhs.isFakeObject() && rhs.isFakeObject() -> {
                     val lhsType = lhs.getFakeType(scope)
                     val rhsType = rhs.getFakeType(scope)
-                    val condition = mkAnd(
+                    mkAnd(
                         lhsType.boolTypeExpr eq rhsType.boolTypeExpr,
                         lhsType.fpTypeExpr eq rhsType.fpTypeExpr,
                         // TODO support type equality
                         lhsType.refTypeExpr eq rhsType.refTypeExpr
                     )
-                    scope.assert(condition)
                 }
 
                 lhs.isFakeObject() -> {
                     val lhsType = lhs.getFakeType(scope)
-                    val condition = when (rhs.sort) {
+                    when (rhs.sort) {
                         boolSort -> lhsType.boolTypeExpr
                         fp64Sort -> lhsType.fpTypeExpr
                         // TODO support type equality
                         addressSort -> lhsType.refTypeExpr
                         else -> error("Unsupported sort ${rhs.sort}")
                     }
-                    scope.assert(condition)
                 }
 
                 rhs.isFakeObject() -> {
                     val rhsType = rhs.getFakeType(scope)
-                    val condition = when (lhs.sort) {
+                    when (lhs.sort) {
                         boolSort -> rhsType.boolTypeExpr
                         fp64Sort -> rhsType.fpTypeExpr
                         // TODO support type equality
                         addressSort -> rhsType.refTypeExpr
                         else -> error("Unsupported sort ${lhs.sort}")
                     }
-                    scope.assert(condition)
                 }
+                else -> error("Should not be called")
             }
 
             return with(Eq) {
-                resolveFakeObject(lhs, rhs, scope)
+                mkAnd(
+                    typeConstraint,
+                    resolveFakeObject(lhs, rhs, scope)
```

Comment:
It is better to extract the "loose equality" result into a separate variable and conjunct it with `typeConstraints` afterwards. This would better reflect the delegation semantics of StrictEq.